File: backend/lib/adapters/evm.js
export async function verifyEvmTx(
  network,
  txHash,
  expectedRecipient,
  expectedAsset,
  expectedAmount,
  tokenAddress,
  tokenDecimals = 18
) {
  const rpcUrl = network.rpcUrl;

  try {
    const txResponse = await fetch(rpcUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 1,
        method: "eth_getTransactionByHash",
        params: [txHash]
      })
    });
    const txResult = await txResponse.json();
    const tx = txResult.result;

    if (!tx) {
      return { success: false, reason: "Transaction not found on chain yet.", pending: true };
    }

    const receiptResponse = await fetch(rpcUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 2,
        method: "eth_getTransactionReceipt",
        params: [txHash]
      })
    });
    const receiptResult = await receiptResponse.json();
    const receipt = receiptResult.result;

    if (!receipt) {
      return { success: false, reason: "Transaction is still pending.", pending: true };
    }

    if (receipt.status !== "0x1") {
      return { success: false, reason: "Transaction failed on-chain." };
    }

    const blockNumberResponse = await fetch(rpcUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 3,
        method: "eth_blockNumber",
        params: []
      })
    });
    const blockNumberResult = await blockNumberResponse.json();
    const currentBlock = parseInt(blockNumberResult.result, 16);
    const txBlock = parseInt(tx.blockNumber, 16);
    const confirmations = currentBlock - txBlock;

    if (confirmations < 1) {
      return { success: false, reason: "Insufficient block confirmations.", pending: true, confirmations };
    }

    const nativeSymbol = network.native?.symbol;
    if (expectedAsset === nativeSymbol) {
      const recipientMatch = tx.to && tx.to.toLowerCase() === expectedRecipient.toLowerCase();
      const nativeDecimals = network.native?.decimals ?? 18;
      const expectedWei = BigInt(Math.floor(parseFloat(expectedAmount) * 10 ** nativeDecimals));
      const actualWei = BigInt(tx.value);

      if (!recipientMatch) {
        return { success: false, reason: `Recipient mismatch. Expected: ${expectedRecipient}, Got: ${tx.to}` };
      }
      if (actualWei < expectedWei) {
        return { success: false, reason: `Native amount mismatch for ${expectedAsset}.` };
      }
    } else if (tokenAddress) {
      const contractMatch = tx.to && tx.to.toLowerCase() === tokenAddress.toLowerCase();
      if (!contractMatch) {
        return { success: false, reason: `Asset contract mismatch. Expected: ${tokenAddress}, Got: ${tx.to}` };
      }

      const data = tx.input;
      if (!data || !data.startsWith("0xa9059cbb")) {
        return { success: false, reason: "Transaction is not an ERC20 transfer call." };
      }

      const recipientHex = "0x" + data.slice(34, 74);
      const valueHex = "0x" + data.slice(74, 138);
      const recipientMatch = recipientHex.toLowerCase() === expectedRecipient.toLowerCase();
      const tokenScale = BigInt(10 ** tokenDecimals);
      const expectedTokens = BigInt(Math.floor(parseFloat(expectedAmount) * Number(tokenScale)));
      const actualTokens = BigInt(valueHex);

      if (!recipientMatch) {
        return { success: false, reason: `Token recipient mismatch. Expected: ${expectedRecipient}` };
      }
      if (actualTokens < expectedTokens) {
        return {
          success: false,
          reason: `Token amount mismatch. Expected: ${expectedAmount} ${expectedAsset}`
        };
      }
    } else {
      return { success: false, reason: "Unsupported asset route." };
    }

    return { success: true, confirmations };
  } catch (err) {
    console.error("EVM verification error:", err);
    return { success: false, reason: `Verification node error: ${err.message}` };
  }
}

File: x402-module/src/adapters/evm.rs
use serde_json::json;

#[derive(Debug)]
pub struct VerifyResult {
    pub success: bool,
    pub reason: Option<String>,
    pub pending: bool,
    pub confirmations: Option<u64>,
}

pub async fn verify_evm_tx(
    rpc_url: &str,
    tx_hash: &str,
    expected_recipient: &str,
    expected_asset: &str,
    expected_amount: &str,
    token_address: Option<&str>,
    native_symbol: &str,
    native_decimals: u32,
    token_decimals: u32,
) -> VerifyResult {
    let client = reqwest::Client::new();

    let tx: Option<serde_json::Value> = rpc_call(&client, rpc_url, "eth_getTransactionByHash", json!([tx_hash]))
        .await
        .ok()
        .and_then(|v| v.get("result").cloned())
        .filter(|v| !v.is_null());

    let Some(tx) = tx else {
        return VerifyResult {
            success: false,
            reason: Some("Transaction not found on chain yet.".into()),
            pending: true,
            confirmations: None,
        };
    };

    let receipt: Option<serde_json::Value> =
        rpc_call(&client, rpc_url, "eth_getTransactionReceipt", json!([tx_hash]))
            .await
            .ok()
            .and_then(|v| v.get("result").cloned())
            .filter(|v| !v.is_null());

    let Some(receipt) = receipt else {
        return VerifyResult {
            success: false,
            reason: Some("Transaction is still pending.".into()),
            pending: true,
            confirmations: None,
        };
    };

    if receipt.get("status").and_then(|s| s.as_str()) != Some("0x1") {
        return VerifyResult {
            success: false,
            reason: Some("Transaction failed on-chain.".into()),
            pending: false,
            confirmations: None,
        };
    }

    let current_block = rpc_call(&client, rpc_url, "eth_blockNumber", json!([]))
        .await
        .ok()
        .and_then(|v| v.get("result").and_then(|s| s.as_str()).map(parse_hex_u64))
        .unwrap_or(0);
    let tx_block = tx
        .get("blockNumber")
        .and_then(|s| s.as_str())
        .map(parse_hex_u64)
        .unwrap_or(0);
    let confirmations = current_block.saturating_sub(tx_block);

    if confirmations < 1 {
        return VerifyResult {
            success: false,
            reason: Some("Insufficient block confirmations.".into()),
            pending: true,
            confirmations: Some(confirmations),
        };
    }

    if expected_asset == native_symbol {
        let to = tx.get("to").and_then(|s| s.as_str()).unwrap_or_default();
        if !eq_addr(to, expected_recipient) {
            return fail("Recipient mismatch.");
        }
        let value = tx
            .get("value")
            .and_then(|s| s.as_str())
            .map(parse_hex_u128)
            .unwrap_or(0);
        let expected = amount_to_units(expected_amount, native_decimals);
        if value < expected {
            return fail("Native amount mismatch.");
        }
    } else if let Some(token) = token_address {
        let to = tx.get("to").and_then(|s| s.as_str()).unwrap_or_default();
        if !eq_addr(to, token) {
            return fail("Asset contract mismatch.");
        }
        let input = tx.get("input").and_then(|s| s.as_str()).unwrap_or_default();
        if !input.starts_with("0xa9059cbb") {
            return fail("Transaction is not an ERC20 transfer.");
        }
        let recipient = format!("0x{}", &input[34..74]);
        if !eq_addr(&recipient, expected_recipient) {
            return fail("Token recipient mismatch.");
        }
        let value_hex = format!("0x{}", &input[74..138]);
        let actual = parse_hex_u128(&value_hex);
        let expected = amount_to_units(expected_amount, token_decimals);
        if actual < expected {
            return fail("Token amount mismatch.");
        }
    } else {
        return fail("Unsupported asset route.");
    }

    VerifyResult {
        success: true,
        reason: None,
        pending: false,
        confirmations: Some(confirmations),
    }
}

async fn rpc_call(
    client: &reqwest::Client,
    rpc_url: &str,
    method: &str,
    params: serde_json::Value,
) -> Result<serde_json::Value, reqwest::Error> {
    let body = json!({ "jsonrpc": "2.0", "id": 1, "method": method, "params": params });
    client
        .post(rpc_url)
        .json(&body)
        .send()
        .await?
        .json()
        .await
}

fn parse_hex_u64(hex: &str) -> u64 {
    u64::from_str_radix(hex.trim_start_matches("0x"), 16).unwrap_or(0)
}

fn parse_hex_u128(hex: &str) -> u128 {
    u128::from_str_radix(hex.trim_start_matches("0x"), 16).unwrap_or(0)
}

fn amount_to_units(amount: &str, decimals: u32) -> u128 {
    let scale = 10u128.pow(decimals);
    (amount.parse::<f64>().unwrap_or(0.0) * scale as f64) as u128
}

fn eq_addr(a: &str, b: &str) -> bool {
    a.to_lowercase() == b.to_lowercase()
}

fn fail(reason: &str) -> VerifyResult {
    VerifyResult {
        success: false,
        reason: Some(reason.into()),
        pending: false,
        confirmations: None,
    }
}
